### PR TITLE
Misc

### PR DESCRIPTION
**Pull request by dweindl**
Untangle amici.cpp to prepare for #122 #123 and reduce compilation time
- Slim down amici.cpp: turn some functions into *Data members
- Get rid of compile-time dependencies of amici.cpp on model code

### CODE REVIEW

File: @amimodel/compileC.m [hunk 1]
```diff
@@ -176,35 +80,35 @@ function compileC(this)
         end
     end
     
-    if(this.cfun(1).(this.funs{j}))
-        fprintf(['ffuns | ']);
-        ffuns = cellfun(@(x) fullfile(this.wrap_path,'models',this.modelname,[this.modelname '_' x '.cpp']),this.funs,'UniformOutput',false);
+    
+    funsForRecompile = this.funs(structfun(@(x) logical(x), this.cfun(1)));
+    if(numel(funsForRecompile))
```
Comment (dweindl) [inline @amimodel/compileC.m]: It is subsetted, it contains only the function names that have to be recompiled. We could sum over cfun, yes, but we need the subset afterwards anyways.

File: @amimodel/compileC.m [hunk 2]
```diff
@@ -123,40 +35,32 @@ function compileC(this)
     
     % generate hash for file and append debug string if we have an md5
     % file, check this hash against the contained hash
-    cppsrc = {'symbolic_functions','spline','edata','rdata','udata','tdata'};
+    cppsrc = {'amici', 'symbolic_functions','spline','edata','rdata','udata','tdata', 'amici_interface_matlab', 'amici_misc', 'amici_model_functions'};
```
Comment (dweindl) [inline @amimodel/compileC.m]: No, not anymore. I changed that here, so it is resolved at compile-time. The *wrap.h functions have the same interface. The implementation doesn't matter.

File: @amimodel/compileC.m [hunk 3]
```diff
@@ -4,113 +4,25 @@ function compileC(this)
...
     else
         o_suffix = '.o';
     end
     
+    amiciSourcePath = fullfile(this.wrap_path,'src');
+    modelSourceFolder = fullfile(this.wrap_path,'models',this.modelname);
+
+    [objectsstr, includesstr] = compileAMICIDependencies(this.wrap_path, o_suffix);
+        
+    includesstr = strcat(includesstr,' -I"', modelSourceFolder, '"');
```
Comment (dweindl) [inline @amimodel/compileC.m]: Will do. Didn't even know this existed.

Comment (dweindl) [inline @amimodel/compileC.m]: As far as I can see, this is currently not handled very well. Once the debug libraries have been built, the only way going back to optimized libs is deleting them manually or a change in the version suitesparse/sundials version number. We need to handle that smarter.

Comment (dweindl) [inline @amimodel/compileC.m]: I propose to have the debug and optimized object files in two different directories, and link them depending on `model.debug`.

File: @amimodel/compileC.m (hunk 2 above)
Comment (FFroehlich) [inline @amimodel/compileC.m]: 👍

File: @amimodel/compileC.m (hunk 1 above)
Comment (FFroehlich) [inline @amimodel/compileC.m]: Ah, I missed that you use the logical values to index this.funs.

File: auxiliary/compileAMICIDependencies.m [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+    includesstr = strcat(includesstr,' -I"', fullfile(ssparse_path, 'KLU','Include'), '"');
+    includesstr = strcat(includesstr,' -I"', fullfile(ssparse_path, 'AMD','Include'), '"');
+    includesstr = strcat(includesstr,' -I"', fullfile(ssparse_path, 'COLAMD','Include'), '"');
+    includesstr = strcat(includesstr,' -I"', fullfile(ssparse_path, 'BTF','Include'), '"');
+    includesstr = strcat(includesstr,' -I"', fullfile(ssparse_path, 'SuiteSparse_config'), '"');
+    
+    % compile all the sundials objects if we haven't done so yet
+    sources_sundials = getSourcesSundials();
+    for j=1:length(sources_sundials)
+        if(~exist(fullfile(objectFolder,objects_sundials{j}), 'file') || del_sundials)
```
Comment (FFroehlich) [inline auxiliary/compileAMICIDependencies.m]: can we also compile all the objects in a single call to mex as we do it for the model specific files?

Comment (dweindl) [inline auxiliary/compileAMICIDependencies.m]: Done. Also for src/*.cpp.

**Pull request merged**